### PR TITLE
change Patch to Update

### LINKED ISSUE #160: to support testing use Update rather than Patch to add annotations and labels
fakeClient.Patch() is not implemented. No error is returned when Patch is called and the returned object is messed up. Our tests have been getting back false positives.
Update() has a likely rare window where the object's version is updated during the patch window. For now we will not worry about this.

### PR DESCRIPTION
**Pull request by jeffvance**
fixes #160 
Improves (hopefully) some log messages with better consistency.
Some err messages also updated.

### CODE REVIEW

File: pkg/controller/util.go [hunk 1]
```diff
@@ -138,55 +137,39 @@ func (c *Controller) getSecretName(pvc *v1.PersistentVolumeClaim) (string, error
...
+			pvcClone, err = c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
+			if err == nil {
+				return false, nil // try again
+			}
+			verb = "getting"
+		}
+		glog.V(Vadmin).Infof("error %s pvc \"%s/%s\": %v\n", verb, pvc.Namespace, pvc.Name, err)
+		return false, nil // try again
+	})
+	return pvc, err
```
Comment (copejon) [inline pkg/controller/util.go]: err needs to be `errors.Wrap()`'ed

File: pkg/controller/util.go [hunk 2]
```diff
@@ -138,55 +137,39 @@ func (c *Controller) getSecretName(pvc *v1.PersistentVolumeClaim) (string, error
...
+	// loop a few times in case the cloned pvc is stale
+	err = wait.PollImmediate(time.Second*1, time.Second*5, func() (bool, error) {
+		pvc, err = c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Update(pvcClone)
+		if err == nil {
+			return true, nil // successful update
+		}
+		verb := "updating"
+		if apierrs.IsConflict(err) { // pvc is likely stale
+			glog.V(Vdebug).Infof("pvc \"%s/%s\" is stale, re-getting pvc\n", pvc.Namespace, pvc.Name)
+			pvcClone, err = c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
```
Comment (jeffvance) [inline pkg/controller/util.go]: cannot re-Get the pvc because the changed applied by the caller will be lost.

File: pkg/controller/util.go [hunk 3]
```diff
@@ -138,55 +137,47 @@ func (c *Controller) getSecretName(pvc *v1.PersistentVolumeClaim) (string, error
...
+		if apierrs.IsConflict(err) { // pvc is likely stale
+			glog.V(Vdebug).Infof("pvc %q is stale, re-trying\n", nsName)
+			pvcCopy, err = c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
+			if err == nil {
+				pvcCopy.ObjectMeta = metadata // remember orig metadata
+				return false, nil             // retry update
+			}
+		}
+		glog.V(Vdebug).Infof("%q update/get error: %v\n", nsName, err)
+		return false, nil // retry
```
Comment (jeffvance) [inline pkg/controller/util.go]: bug here: if the Get fails we retry Update with a zero-value pvc.

File: pkg/controller/util.go [hunk 4]
```diff
@@ -138,55 +137,47 @@ func (c *Controller) getSecretName(pvc *v1.PersistentVolumeClaim) (string, error
...
-// Note: Patch() is used instead of Update() to handle version related field changes.
-func (c *Controller) setPVCAnnotation(pvc *v1.PersistentVolumeClaim, key, val string) error {
-	glog.V(Vdebug).Infof("Adding annotation \"%s: %s\" to pvc \"%s/%s\"\n", key, val, pvc.Namespace, pvc.Name)
+	// loop a few times in case the pvc is stale
+	err = wait.PollImmediate(time.Second*1, time.Second*4, func() (bool, error) {
+		updtPvc, err = c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Update(pvcCopy)
+		if err == nil {
+			return true, nil // successful update
+		}
+		if apierrs.IsConflict(err) { // pvc is likely stale
```
Comment (jeffvance) [inline pkg/controller/util.go]: better to not retry at all, not even have a loop here. Let the re-queue logic of the pvc work queue handle the retry.

File: pkg/controller/util.go [hunk 5]
```diff
@@ -138,55 +137,47 @@ func (c *Controller) getSecretName(pvc *v1.PersistentVolumeClaim) (string, error
...
-	}
-	return pvcClone, data, nil
-}
-
-// Sets an annotation `key: val` in the given PVC.
-// Note: Patch() is used instead of Update() to handle version related field changes.
-func (c *Controller) setPVCAnnotation(pvc *v1.PersistentVolumeClaim, key, val string) error {
-	glog.V(Vdebug).Infof("Adding annotation \"%s: %s\" to pvc \"%s/%s\"\n", key, val, pvc.Namespace, pvc.Name)
+	// loop a few times in case the pvc is stale
+	err = wait.PollImmediate(time.Second*1, time.Second*4, func() (bool, error) {
```
Comment (jeffvance) [inline pkg/controller/util.go]: increase 4s to maybe 20s?

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -121,36 +121,38 @@ func (c *Controller) runPVCWorkers() {
...
 	key, shutdown := c.podQueue.Get()
 	if shutdown {
 		return false
 	}
 	defer c.podQueue.Done(key)
+
 	pod, err := c.podFromKey(key)
 	if err != nil {
-		c.forgetKey(key, fmt.Sprintf("Unable to get pod object: %v", err))
+		c.forgetKey(key, fmt.Sprintf("ProcessNextPodItem: unable to get pod from key %q: %v", key, err))
```
Comment (copejon) [inline pkg/controller/controller.go]: One error case is that `key` is not parsable as a string, so `%q` will print out garbage.

File: pkg/controller/util.go [hunk 7]
```diff
@@ -138,55 +137,67 @@ func (c *Controller) getSecretName(pvc *v1.PersistentVolumeClaim) (string, error
...
-	if err != nil {
-		return errors.Wrap(err, "error creating pvc patch")
-	}
-	_, err = c.clientset.CoreV1().PersistentVolumeClaims(pvc.Namespace).Patch(pvc.Name, types.StrategicMergePatchType, patch)
-	if err != nil {
-		return errors.Wrapf(err, "error patching pvc %q", pvc.Name)
-	}
-	return nil
-}
+// Update and return a copy of the passed-in pvc. Only one of the annotation or label maps is requred though
```
Comment (copejon) [inline pkg/controller/util.go]: typo in "required"

**Pull request merged**